Inspect dataset structure and propose way to use it for our evaluation #14

Rut-Vahab opened the pull request:
# Dataset Inspection & Preparation

## 1. Original Structure
- Files are provided in **segmentation** format: images + masks.  
- **Labels** are given as `LabelID` in a CSV file.  
- **No** predefined split into train/val/test exists.  

## 2. Label Mapping
- A mapping was created between `LabelID` and class names (from `datasets.py`).  
- Found **85 active classes** out of 101.  
- Removed unused classes to simplify the workflow.  

## 3. Conversion to Classification Structure
- A new structure was created in **classification-style** format:  
dataset/classifier_data/
├── 0/
│ ├── image_1.png
│ ├── image_2.png
├── 1/
│ ├── image_1.png
│ ├── image_2.png

- Each class has its own folder with its examples.  

## 4. Dataset Splits
- Since no predefined splits exist → we created a custom **train/val** split.  
- The split is applied during preparation without modifying the original dataset.  

## 5. Preparation Script
Added steps to `preprocess_data.sh`:  
1. Download and prepare the original dataset.  
2. Convert to class-based (classification) structure.  

---

## ✅ Acceptance Criteria
- Full documentation of dataset structure.  
- Unused classes removed (85 remained).  
- Dataset converted to classification-friendly structure.  
- Train/val split added.  
- Preparation script (`preprocess_data.sh`) generates the complete format automatically.  
- Decided to use **Macro-F1, Balanced Accuracy, Multiclass MCC** to handle class imbalance.  

---

**Dataset is ready for use in the evaluation pipeline.**

chani0343 committed:
Inspect dataset structure, add conversion and statistics scripts, update docs

Rut-Vahab committed:
Add dataset inspection and classification conversion details for evaluation #14

chani0343 committed:
Merge branch 'main' into inspect-dataset-structure-14

chani0343 committed:
Complete dataset structure documentation and evaluation approach

- Finalize comprehensive dataset preparation documentation
- Document successful conversion to classification structure (83 classes)
- Add implementation results with working inference pipeline
- Include sample predictions and performance metrics
- Complete all acceptance criteria for issue #14

chani0343 committed:
Merge branch 'inspect-dataset-structure-14' of https://github.com/KamaTechOrg/AgStream into inspect-dataset-structure-14
:wq

Rut-Vahab committed:
Resolve merge conflicts in dataset_preparation.md

Rut-Vahab committed:
docs: update dataset preparation and classification pipeline documentation

r83575 commented on scripts/preprocess_data.sh:
```diff
@@ -0,0 +1,12 @@
+python dataset_preparation/dataset_prep.py --data_root $DATA_ROOT
+
+echo "=== Step 2: Mapping dataset to classification format ==="
+python dataset_preparation/map_dataset.py --dataset_target CropsOrWeed9
```
First of all, you don’t need to map it to the CropsOrWeed9 dataset; we use the CropsAndWeed dataset, and the mapping process is not required because it’s already mapped.
In addition, currently it’s problematic to convert it into a classification format, because each image may include multiple types of plants.

lyuzinmaxim replied:
Are these 2 steps enough to run evaluation on the dataset?

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+
+## Conversion Process
+
+1. **Source:** Semantic masks from `labelIds/CropAndWeed/`  
```
Both masks and bounding boxes have the category we need. I suppose it might be more easy to work with bounding boxes than with masks

lyuzinmaxim commented on docs/dataset_preparation.md:
````diff
@@ -1,45 +1,175 @@
+
+### Install Requirements
+```bash
+pip install requests tqdm opencv-python torch torchvision
````
Since we are working from inside the docker environment, we don't need these steps. If you need to install smth, just add this to `requirements.txt`

lyuzinmaxim commented on docs/dataset_preparation.md:
````diff
@@ -1,45 +1,175 @@
+
+### Prepare Dataset
+```bash
+bash prepare_dataset.sh
````
Once you are sure that it's working from within the docker env, please add a note that this command should be executed from the inside of the container

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+
+---
+
+## Split Strategy ✅
```
Let's not split it now - anyway we have zero idea how was the data split for training.

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+## Challenges & Solutions
+
+### 1. Label Mapping  
+- **Problem:** Pixel mask → single label  
```
already mentioned that it might be easier to use bboxes for that

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+
+### 2. Class Imbalance  
+- **Problem:** Few samples in some classes  
+- **Solution:** Filtered → 83 balanced classes  
```
Please specify how exactly did you filter, and what are the removed categories (and why removed)

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+- **Problem:** Few samples in some classes  
+- **Solution:** Filtered → 83 balanced classes  
+
+### 3. Data Loading Performance  
```
Let's remove this point - it's clear from the stage

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+
+## Evaluation Strategy
+
+- **Metrics:** Accuracy, Loss (TensorBoard)  
```
Why tensorboard? We don't know that

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+
+- **Metrics:** Accuracy, Loss (TensorBoard)  
+- **Validation:** 20% held-out data  
+- **Models tested:** MobileNetV2, ResNet18, ShuffleNet  
```
What shufflenet?

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+- **Validation:** 20% held-out data  
+- **Models tested:** MobileNetV2, ResNet18, ShuffleNet  
+
+**Future Improvements:**  
```
Let's not mention it in here

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+
+## Results ✅
+
+### Training
```
We don't train the model, please remove it

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+
+---
+
+## Results ✅
```
I'd remove this points "Results"

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,175 @@
+
+---
+
+## Summary
```
This section to remove as well

lyuzinmaxim commented on scripts/preprocess_data.sh:
```diff
@@ -0,0 +1,12 @@
+echo "=== Step 2: Mapping dataset to classification format ==="
+python dataset_preparation/map_dataset.py --dataset_target CropsOrWeed9
+
+echo "=== Step 3: Done. Dataset ready at: $DATA_ROOT/classifier_data ==="
```
it's not a step; it's an outcome

chani0343 commented:
Hi ,

Thank you for all the feedback. I've updated the documentation based on your comments:

**Changes made:**
- ✅ Removed all references to CropsOrWeed9 mapping (not needed)
- ✅ Emphasized bounding boxes over semantic masks (easier to work with)
- ✅ Updated Docker environment instructions (execute inside container)
- ✅ Modified split strategy (use dataset as-is for evaluation)
- ✅ Focused on evaluation approach rather than training

**Key points addressed:**
- No mapping step needed - dataset is already properly labeled
- Bounding boxes are the recommended approach for label extraction
- Multi-label challenge documented with solution options
- Docker-first approach for setup instructions

The documentation now clearly focuses on dataset inspection and evaluation strategy as required by the ticket.

Ready for marge!

chani0343 committed:
docs: finalize dataset inspection and structure documentation

- Complete dataset structure analysis with native format documentation
- Identify bounding boxes as recommended approach for label extraction
- Address multi-label challenge with 3 solution options
- Define evaluation strategy without predetermined splits
- Update setup instructions for Docker environment
- Fulfill all DoD requirements for dataset structure inspection

Resolves feedback from code review:
- Use bounding boxes instead of semantic masks
- Keep original dataset structure (no conversion needed)
- Focus on evaluation approach rather than training pipeline

chani0343 committed:
fix: resolve black formatting issues for linting

- Apply black formatting to analyze_dataset.py
- Apply black formatting to dataset_preparation/dataset_prep.py
- Ensure all Python files pass CI linting checks

chani0343 committed:
Complete dataset inspection and structure documentation

- Document dataset native format (images, bboxes, labelIds, params)
- Identify bounding boxes as recommended approach for labels
- Address multi-label challenge with solution options
- Define evaluation strategy without predetermined splits
- Update setup instructions for Docker environment
- Apply black formatting to all Python files
- Fulfill all DoD requirements for dataset structure inspection

r83575 commented on docs/dataset_preparation.md:
```diff
@@ -41,135 +41,104 @@ dataset/
 
-**Decision Made:** After analysis, we converted to a **classification structure**.  
+### Dataset Splits
+**Answer:** The original dataset does **not** provide predefined train/val/test splits. We need to create them ourselves.
```
There is no question to answer here.
Please just explain it.

Rut-Vahab replied:
Great, we fixed it.

r83575 commented on docs/dataset_preparation.md:
```diff
@@ -41,135 +41,104 @@ dataset/
-- Split: 80/20 (train/val) with PyTorch `random_split`  
+### Approach
+- Extract labels from bounding box CSV files
+- Assign single label per image (dominant class by bbox area or count)
```
The strategy is been changed, Please address it.

Rut-Vahab committed:
Update dataset_preparation.md

Rut-Vahab committed:
Update dataset_preparation.md

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,137 @@
+
+---
+
+## Next Steps
```
I wouldn't add the information about the next steps to README, since for that we have our board.

lyuzinmaxim commented on docs/dataset_preparation.md:
```diff
@@ -1,45 +1,137 @@
+
+---
+
+## Summary
```
Also I wouldn't add the "summary" part since it's selfexplanatory

lyuzinmaxim commented on dataset_preparation/dataset_prep.py:
```diff
@@ -21,6 +21,9 @@ def parse_arguments():
 
 
 def setup(data_root: str):
+    import urllib3
```
What is the purpose of it?

chani0343 replied:
I needed to download the data and I had an SSL problem and I asked GPT and he told me to add this line and that solved my problem
We want to move on to the next task

lyuzinmaxim replied:
since we are running this function each time we're running the script, let's move the import to the top of the file

Rut-Vahab replied:
Ok thanks we'll move it..
Can we merge this now?

Rut-Vahab committed:
Clean up dataset preparation doc: removed Next Steps and Summary

Rut-Vahab committed:
Update dataset_prep.py

Rut-Vahab committed:
Dataset inspection & preparation documentation
